llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,245 @@
+                             const BundledSCEV &BunldedSCEV, int MinIterCount,
+                             AssumptionCache &AC) {
+  Instruction *HeaderTerminator = LoopHeader.getTerminator();
+
+  // LoopRotate uses SimplifyQuery to determine, if a Branch is conditional or
+  // not. SimplifyQuery can only take an Assumption into account, if it is
+  // before the to-be-evaluated Compare Instruction. Thus assumptions have to be
+  // inserted before the LoopCmpInstr.
+  IRBuilder<> Builder(dyn_cast<Instruction>(&LoopCmpInstr));
+
+  SCEVExpander Expander(SE, LoopHeader.getModule()->getDataLayout(),
+                        "expanded");
+
+  // Get Minimum Iteration Value Expression
+  const SCEV *MinIterSCEV = BunldedSCEV.AddRecExpr->evaluateAtIteration(
+      SE.getConstant(APInt(32, MinIterCount)), SE);
+
+  // If the MinIterSCEV is not a constant, but based on an expression,
+  // the no signed/unsigned overflow flags have to be added, so that the
+  // expression can be simplified in later passes
+  SCEVCouldNotCompute::NoWrapFlags NWF = BunldedSCEV.AddRecExpr->getNoWrapFlags(
+      static_cast<SCEVCouldNotCompute::NoWrapFlags>(SCEV::FlagNUW |
+                                                    SCEV::FlagNSW));
+  auto *AE = dyn_cast<SCEVAddExpr>(MinIterSCEV);
+  if (AE && NWF) {
+    MinIterSCEV = SE.getAddExpr(AE->getOperand(0), AE->getOperand(1), NWF);
+  }
+
+  // Expansion of MinIterSCEV will result in                                   \
+  // InitialValue + StepSize * Minimum Iteration Count
+  if (!Expander.isSafeToExpand(MinIterSCEV)) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: Cannot Expand "
+                         "Minimum Iteration Scalar Evolution"
+                      << *MinIterSCEV << "\n");
+    return;
+  }
+  Type *CmpType = LoopCmpInstr.getOperand(0)->getType();
+  Value *LoopAssertValue =
+      Expander.expandCodeFor(MinIterSCEV, CmpType, HeaderTerminator);
+  LLVM_DEBUG(dbgs() << "Loop Iteration Count        : " << MinIterCount << "\n"
+                    << "MinIterValue                : " << *LoopAssertValue
+                    << "\n");
+
+  Value *LHS = BunldedSCEV.IVOperandIndex == 0 ? LoopAssertValue
+                                               : LoopCmpInstr.getOperand(0);
+  Value *RHS = BunldedSCEV.IVOperandIndex == 0 ? LoopCmpInstr.getOperand(1)
+                                               : LoopAssertValue;
+  Value *Cmp = Builder.CreateICmp(LoopCmpInstr.getPredicate(), LHS, RHS);
+
+  // Insert Assumption
+  CallInst *Assumption = Builder.CreateAssumption(Cmp);
+  AC.registerAssumption(dyn_cast<AssumeInst>(Assumption));
+  LLVM_DEBUG(dbgs() << "With Comparator             :" << *Cmp << "\n"
+                    << "Assume                      :" << *Assumption << "\n");
+}
+
+/// Get Minimum Iteration Count of the CurrentLoop if it exists and is greater
+/// than 0
+std::optional<int64_t> getMinIterCount(Loop &CurrentLoop) {
```

Comment:
I think we should rename this to `getMinExitCount()`, because that is really what we are computing. I.e. "the number of times a backedge is taken before the exit is taken".
 - For non-rotated loops, our min_tripcount metadata (which represents the "body entry count") is the same as the backedge taken count because the backedge is un-conditional.
 - For rotated loops, the backedge will be taken at least `min_tripcount - 1` times.

TLDR: I agree with the numbers, but I'd like a bit more details about how they come from. Our idea of a "C trip count" is different from the `llvm::Loop` trip count (which is just the number of times the header is entered). Naming this function `getMinExitCount()` would already help a lot in getting closer to LLVM terminology.

Reply:
that makes sense